tests/unit/test_models.py

```diff
@@ -364,6 +533,116 @@ def test_summarize(self):
             "Total ETH Funds needed: 123.4568\nTotal COW Funds needed: 10000000.0000\n",
         )
 
+    def test_try_redirect(self):
+        """
+        Test demonstrates that try_redirect works as expected for our use case.
+        However, it also demonstrates how bad it is to pass in an unstructured hashmap
+        that expects the keys to be equal the solver field of its values!
+        TODO - fix this strange error prone issue!
+        """
+        dummy_print_store = PrintStore()
+        receiver = Address.from_int(1)
+        redirect = Address.from_int(2)
+        # Try redirect elsewhere
+        t1 = Transfer(token=None, amount_wei=1, receiver=receiver)
+        vouch_forward = Vouch(
+            bonding_pool=Address.zero(), reward_target=redirect, solver=receiver
+        )
+        t1.try_redirect({vouch_forward.solver: vouch_forward}, dummy_print_store)
+        self.assertEqual(t1.recipient, redirect)
+        self.assertEqual(
+            t1,
+            Transfer(token=None, amount_wei=1, receiver=receiver, redirect_to=redirect),
+        )
+
+        vouch_reverse = Vouch(
+            bonding_pool=Address.zero(), reward_target=receiver, solver=redirect
+        )
+        # Redirect back!
+        t1.try_redirect({vouch_reverse.solver: vouch_reverse}, dummy_print_store)
+        self.assertEqual(t1.recipient, receiver)
+        self.assertEqual(
+            t1,
+            Transfer(token=None, amount_wei=1, receiver=receiver, redirect_to=receiver),
+        )
+        # no action redirect.
+        another_address = Address.from_int(5)
+        t2 = Transfer(token=None, amount_wei=1, receiver=another_address)
+        disjoint_redirect_map = {
+            vouch_forward.solver: vouch_forward,
+            vouch_reverse.solver: vouch_reverse,
+        }
```

Comment:
Is this a valid scenario in our domain? Doesn't this mean we are trying to pay out a solver that isn't vouch for? Shouldn't this cause an error?

Reply:
Technically yes because the voucher registry contract has absolutely no constraints.. it's just a function "vouch" that emits events. I have never actually considered this scenario in it vouch query unit tests, but my guess would be that this could actually happen.

Knowing that now, I don't believe it would cause any unusual issue with our current payout script, because we only ever apply the redirect "operator" once. However this would be a problem if redirects were applied twice.

Reply:
I believe any changes related to this should be made in a separate PR -- and that the program is logically sounds with or without weirdness being a thing.